### ISSUE TITLE
Drop down buttons require two clicks to open

### ISSUE THREAD

**Issue by thx1200** [YAFNET/YAFNET-DNN#56]
**Describe the bug**
Drop down buttons require two clicks to open

**To Reproduce**
1. Click on any drop down menu, like the admin settings
2. Nothing happens.
3. Click again and it opens.

**Expected behavior**
Menu should open on first click.

**Screenshots**
Please see this video: https://www.youtube.com/watch?v=STB3VDYem74

**System (please complete the following information):**
 - YAF Version: 2.3.1.04
 
**Affected browser**
At least Chrome

**Additional context**
I tried several different themes because I wasn't sure this was theme related.  It happens on all themes.  I switched back to default theme for the video, so it looks a bit off, but I figure this is the best way to show.

**Comment by w8tcha** [comment]
If I remember this issue is caused when the js file for the bootstrap extension is loaded twice.

I guess you dnn installation skin also losses the js file?

**Comment by thx1200** [comment]
Just checking, the skin does load the js file and dnn inlines it apparently.  Can I tell yaf to not load it somehow?

**Comment by w8tcha** [comment]
At the moment it is not possible. i suggest we use Bootstrap as [Javascript Library extension](http://www.dnnsoftware.com/wiki/javascript-libraries) 

then you would need to register bootstrap in your skin with ...
`<dnn:JavaScriptLibraryInclude runat="server" Name="Bootstrap" />`

**Comment by thx1200** [comment]
I'll look into modifying the skin like that.  So when doing it as a library, Yaf knows not to load its own bootstrap?  I guess i'm confused a bit on how it's getting loaded twice.

**Comment by w8tcha** [comment]
> I'll look into modifying the skin like that. So when doing it as a library, Yaf knows not to load its own bootstrap? I guess i'm confused a bit on how it's getting loaded twice.

In that case dnn would take care of it, and would prevent that a script is loaded more then one time

**Comment by thx1200** [comment]
Thanks, I'll play around with that tonight and report back if it fixes the issue.  Appreciate the guidance.

**Comment by thx1200** [comment]
I got the skin to load successfully with the bootsrtap module.  I also removed all trace of the bootstrap.js files from the skins folder.  The problem still persists.

Looking at the source code, I only see one attempt to load the bootstrap js file.

This line, which is generated from DNN.
<script src="/Resources/libraries/bootstrap/04_04_01/bootstrap.min.js?cdv=181" type="text/javascript"></script>

I do see that jquery is loaded a few different times across several different modules.  Could that be an issue?

**Comment by thx1200** [comment]
OK through trial and error I found a way to make this work.  Apparently with the previous skin settings plus the javascript library method, popper and bootstrap were loading at the bottom of the page.  I did some checking and this seems to be common.  Anyway, I put static references to the js files at the top of the page and everything started working.  The thing is, there's no way to force dnn:JavaScriptLibraryInclude to tell it to load at the top because the module you install specifies load location.  So, while I have a workaround, it's not ideal.

Outside DNN, does YAF load bootstrap at the top or bottom of the page?  I am not really understanding why that makes a difference.  But it seems to.

**Comment by w8tcha** [comment]
Yaf needs to loads the scripts in the header.

Is this workaround actually working?

**Comment by thx1200** [comment]
The workaround works, but it's a hack.  I am posting to DNN forums to see if it's possible to influence where the js files are loaded.  It doesn't matter where dnn:JavaScriptLibraryInclude is positioned in the skin file, DNN will always load it at the bottom because that's apparently where it think it is supposed to load.

I suppose I can go back to static links, but I liked the idea of dnn:JavaScriptLibraryInclude since it prevents double loading and such.

**Comment by w8tcha** [comment]
Sorry but i don't understand. To reproduce the issue what workaround are you using right now?

**Comment by thx1200** [comment]
The workaround is just putting direct js references to the libraries at the top of the skin.

**Comment by w8tcha** [comment]
Okay, but why not use he suggested method? Of course I need to update yaf first to also use the function.

**Comment by thx1200** [comment]
What suggested method?  If you mean dnn:JavaScriptLibraryInclude runat="server" Name="Bootstrap", it's because that library loader puts the scripts at the bottom of the page, no matter where the code is placed in the skin file.

**Comment by w8tcha** [comment]
And that is a problem for your skin?

**Comment by w8tcha** [comment]
I Will release an updated version of yaf that uses javascriptLibaryInclude. Including the need Bootstrap Libary. atleast in my dnn installation it works

**Comment by thx1200** [comment]
Well in the skin I can't specify where the js loads, so it loads below yaf, which makes it not work.  So that's great you are able to add that to yaf!  I'll just keep the workaround in place until the release.  Thanks!

**Comment by w8tcha** [comment]
Okay one update it is possible to load the javascript libary in the head. it need to be defined in the package file...

[bootstrap-bundle_4.4.1.zip](https://github.com/YAFNET/YAFNET-DNN/files/4604881/bootstrap-bundle_4.4.1.zip)

you can try it out .. put this in your skin

`<dnn:JavaScriptLibraryInclude runat="server" Name="bootstrap-bundle" />`

**Comment by thx1200** [comment]
I tried that, no luck.  I spent a lot of time this evening trying to figure out what was different in the working and non-working states.  It seems to be a combination of two things, which made this difficult.  One is that the scripts had to be at the top.  Your bundle package, or modifying the original popper/bootstrap packages did that.  The second part, which only seemed to exist when the script was referenced at the bottom, was an odd "type" on the script tag of type="fa5effb49f2b189f724a0c69-text/javascript"

Take that away, it doesn't work, double click issue.  Put it there, no double click issue.  The only way to get both at the same time was to manually put in a reference and avoid the dnn library loaders altogether.  I really don't understand.  But it works, so this is fine for now.  I ended up just putting static references to the CDN bootstrap with that magic type in the library reference and the site works.  I cannot explain it.
